AF-2471: Create Dashbuilder Runtimes Module

jesuino opened the pull request:
Create the module Dashbuilder Runtime.

Dashbuilder Runtime is a standalone application used to run Dashboards exported from Business Central.

Please merge with https://github.com/kiegroup/droolsjbpm-build-bootstrap/pull/1316

ederign commented on dashbuilder/dashbuilder-runtime/src/main/java/org/dashbuilder/backend/resources/RuntimeJaxAppApp.java:
```diff
@@ -0,0 +1,25 @@
+import javax.ws.rs.ApplicationPath;
+import javax.ws.rs.core.Application;
+
+@ApplicationPath("/rest")
```
Why this is needed?

jesuino replied:
We use this to expose REST services. Currently we have one REST service to upload files, I am thinking about making the call to RuntimeModelService use REST instead RPC

ederign commented on dashbuilder/dashbuilder-runtime/src/main/java/org/dashbuilder/client/plugins/RuntimePerspectivePluginManager.java:
```diff
@@ -0,0 +1,131 @@
+
+    @Override
+    public void buildPerspectiveWidget(String perspectiveName, LayoutTemplateContext layoutCtx, ParameterizedCommand<IsWidget> afterBuild, ParameterizedCommand<LayoutRecursionIssue> onInfiniteRecursion) {
+        // LAYOUT RECURSION SHOULD BE HANDLED HERE
```
Could you please clarify?

jesuino replied:
This layout recursion should not be a problem after authoring, but I left a comment to remember where to fix if we face it.

ederign commented on dashbuilder/dashbuilder-runtime/src/main/java/org/dashbuilder/shared/model/DashbuilderRuntimeMode.java:
```diff
@@ -0,0 +1,40 @@
+    /**
+     * Users can import how many dashboards they want and reopen them later.
+     */
+    MULTIPLE_IMPORT;
```
William, I think we should remove this option at this stage and add it later if necessary.

ederign commented:
@caponetto could you please take a look at this when you have a chance?

@jesuino could you please explain what we are trying to achieve via chat/call to Guilherme?

jesuino commented on uberfire-extensions/uberfire-runtime-plugins/uberfire-runtime-plugins-client/src/main/java/org/uberfire/ext/plugin/client/perspective/editor/generator/PerspectiveEditorActivity.java:
```diff
@@ -95,7 +95,7 @@ public PerspectiveDefinition buildPerspective() {
     }
 
-    private String getDefaultPanelType() {
+    protected String getDefaultPanelType() {
```
Yes, we had to change the PanelType specifically for the Runtime. We didn't want to possibly break generated panels in Business Central. See the class RuntimePerspectiveEditorActivity

jesuino commented on dashbuilder/dashbuilder-runtime/src/main/java/org/dashbuilder/backend/RuntimeOptions.java:
```diff
@@ -0,0 +1,128 @@
+
+    @PostConstruct
+    public void init() {
+        String multipleImportStr = System.getProperty(DASHBUILDER_RUNTIME_MULTIPLE_IMPORT, "false");
+        String allowExternalStr = System.getProperty(ALLOW_EXTERNAL_FILE_REGISTER_PROP, "true");
```
Thanks, you also reminded me that I should use default false for allowExternalStr!

jesuino commented on dashbuilder/dashbuilder-runtime/src/main/java/org/dashbuilder/backend/remote/services/RuntimeModelServiceImpl.java:
```diff
@@ -0,0 +1,77 @@
+        }
+
+        // if it is an existing file
+        Optional<String> modelPath = runtimeOptions.modelPath(exportId);
+        if (modelPath.isPresent()) {
+            return importModelRegistry.registerFile(modelPath.get());
+        }
```
In this I think that ifPresent would not help us because I need to return the optional value. I thought about chaining all optional calls, but I think the code would be more unclear. The If's sequence is helpful to expose what would be the order to load stuff:

* No modelId is provided, so we try to return the only one that is registered;
* Next we try to load some existing file with the given model Id;
* Finally we try to load external files (only if it is enabled to load external files)

if we do not allow multiple imports than we simple return the single one that is potentially loaded.

jesuino commented on dashbuilder/dashbuilder-runtime/src/main/java/org/dashbuilder/backend/RuntimeOptions.java:
```diff
@@ -0,0 +1,128 @@
+
+    private static final String DASHBUILDER_RUNTIME_MULTIPLE_IMPORT = "dashbuilder.runtime.multiple";
+
+    private static final int DEFAULT_UPLOAD_SIZE = 96 * 1024;
```
It should be in KB - export models without big CSVs inside it would always be <100kb

jesuino commented on dashbuilder/dashbuilder-client/dashbuilder-navigation-client/src/main/java/org/dashbuilder/client/navigation/plugin/PerspectivePluginManagerImpl.java:
```diff
@@ -0,0 +1,241 @@
+
+    @Override
+    public void buildPerspectiveWidget(String perspectiveName, LayoutTemplateContext layoutCtx, ParameterizedCommand<IsWidget> afterBuild, ParameterizedCommand<LayoutRecursionIssue> onInfiniteRecursion) {
+        Plugin plugin = pluginMap.get(perspectiveName);
```
I have to agree but this code was a refactor from PerspectivePluginManager. We had to re-implement the behavior of this class in Runtime, so I had to extract and interface out of it and add a specific implementation in Runtime! Hence why I didn't modify any of the original code to avoid breaking stuff in other webapps


[1] https://github.com/kiegroup/appformer/blob/master/dashbuilder/dashbuilder-client/dashbuilder-navigation-client/src/main/java/org/dashbuilder/client/navigation/plugin/PerspectivePluginManager.java#L147

jesuino commented on dashbuilder/dashbuilder-client/dashbuilder-navigation-client/src/main/java/org/dashbuilder/client/navigation/plugin/PerspectivePluginManagerImpl.java:
```diff
@@ -0,0 +1,241 @@
+        NavWorkbenchCtx ctx = NavWorkbenchCtx.perspective(pluginName);
+        NavTree navTree = navigationManager.getNavTree();
+        List<NavItem> itemsToDelete = navTree.searchItems(ctx);
+        for (NavItem item : itemsToDelete) {
```
See this comment -> https://github.com/kiegroup/appformer/pull/956#discussion_r426657494

jesuino commented on dashbuilder/dashbuilder-client/dashbuilder-navigation-client/src/main/java/org/dashbuilder/client/navigation/plugin/PerspectivePluginManagerImpl.java:
```diff
@@ -0,0 +1,241 @@
+            NavWorkbenchCtx ctx = NavWorkbenchCtx.perspective(event.getOldPluginName());
+            NavWorkbenchCtx newCtx = NavWorkbenchCtx.perspective(event.getPlugin().getName());
+            List<NavItem> itemsToRename = navigationManager.getNavTree().searchItems(ctx);
+            for (NavItem navItem : itemsToRename) {
```
See this comment -> https://github.com/kiegroup/appformer/pull/956#discussion_r426657494

jesuino commented on dashbuilder/dashbuilder-client/dashbuilder-navigation-client/src/main/java/org/dashbuilder/client/navigation/plugin/PerspectivePluginManagerImpl.java:
```diff
@@ -0,0 +1,241 @@
+
+    // Sync up both the internals plugin & widget registry
+
+    public void onPlugInAdded(@Observes final PluginAdded event) {
```
See this comment -> https://github.com/kiegroup/appformer/pull/956#discussion_r426657494

jesuino commented on dashbuilder/dashbuilder-client/dashbuilder-navigation-client/src/main/java/org/dashbuilder/client/navigation/plugin/PerspectivePluginManagerImpl.java:
```diff
@@ -0,0 +1,241 @@
+    @Override
+    public boolean isRuntimePerspective(Plugin plugin) {
+        ClientResourceType type = clientTypeRegistry.resolve(plugin.getPath());
+        return type != null && type instanceof PerspectiveLayoutPluginResourceType;
```
See this comment -> https://github.com/kiegroup/appformer/pull/956#discussion_r426657494

jesuino commented on dashbuilder/dashbuilder-runtime/src/main/java/org/dashbuilder/client/plugins/RuntimePerspectivePluginManager.java:
```diff
@@ -0,0 +1,131 @@
+                                                              PluginType.PERSPECTIVE,
+                                                              null))
+                                        .collect(Collectors.toList());
+        callback.execute(plugins);
```
Yes, otherwise it would make the UI stuck in some of the navigation drag components

jesuino commented on dashbuilder/dashbuilder-runtime/pom.xml:
```diff
@@ -0,0 +1,893 @@
+		<relativePath>../pom.xml</relativePath>
+	</parent>
+
+	<artifactId>dashbuilder-runtime</artifactId>
```
Thanks! 

Created PR https://github.com/kiegroup/droolsjbpm-build-bootstrap/pull/1316

jesuino commented on dashbuilder/dashbuilder-runtime/src/main/java/org/dashbuilder/backend/services/impl/RuntimeModelParserImpl.java:
```diff
@@ -0,0 +1,138 @@
+    }
+
+    private DataSetContent retrieveDataSetContent(ZipEntry entry, ZipInputStream zis) {
+        String fileName = entry.getName().split("/")[3];
+        String[] nameParts = fileName.split("\\.");
+        String id = nameParts[0];
+        String ext = nameParts[1];
```
The path is retrieved from a file supposedly exported from BC or Dashbuilder. In these cases the directory structure would have to follow zip entry specification, see in [1]:

```
       4.4.17.1 The name of the file, with optional relative path.
       The path stored MUST NOT contain a drive or
       device letter, or a leading slash.  All slashes
       MUST be forward slashes '/' as opposed to
       backwards slashes '\' for compatibility with Amiga
       and UNIX file systems etc.  If input came from standard
       input, there is no file name field.  
```

So it should use forward slash. 

However, DataTransferServices should be tested on Windows environment. But bear in mind that unfortunately it is broken on Windows, see [2]. Once [2] is solved we should revisit this topic to make sure that exported files in Windows will work with runtime, hence I created [3].


[1] https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT
[2] https://issues.redhat.com/browse/RHPAM-2751
[3] https://issues.redhat.com/browse/AF-2520

jesuino commented:
Hello @caponetto  - Thanks again for your great review.

I just made all the requests changes with some small exceptions. Would you please verify it?

Hello @barboras7,  This is the PR for the Dashbuilder Standalone bundle (here we call it dashbuilder-runtime) Would you please let us know who will verify it?

Thanks!

jesuino commented:
Jenkins execute full downstream build

caponetto commented on dashbuilder/dashbuilder-runtime/src/main/java/org/dashbuilder/client/screens/view/UploadDashboardsScreenView.java:
```diff
@@ -0,0 +1,92 @@
+package org.dashbuilder.client.screens.view;
```
```suggestion
/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.dashbuilder.client.screens.view;
```

caponetto commented on dashbuilder/dashbuilder-runtime/src/main/java/org/dashbuilder/client/screens/view/RuntimeScreenView.java:
```diff
@@ -0,0 +1,66 @@
+package org.dashbuilder.client.screens.view;
```
```suggestion
/*
 * Copyright 2020 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.dashbuilder.client.screens.view;
```

caponetto commented on dashbuilder/dashbuilder-runtime/src/main/java/org/dashbuilder/client/screens/view/RuntimeScreenView.html:
```diff
@@ -0,0 +1,31 @@
+            <div
+                class="container-fluid runtime-container-fluid--blank-slate">
+                <div class="blank-slate-pf row">
+                    <h1 data-i18n-key="title">Welcome to Dashbuilder Runtimes</h1>
```
```suggestion
                    <h1 data-i18n-key="title"></h1>
```
Since you have an i18n-key binding, you don't need the hardcoded string.

caponetto commented on dashbuilder/dashbuilder-runtime/src/main/java/org/dashbuilder/client/screens/view/RuntimeScreenView.html:
```diff
@@ -0,0 +1,31 @@
+                <div class="blank-slate-pf row">
+                    <h1 data-i18n-key="title">Welcome to Dashbuilder Runtimes</h1>
+
+                    <p data-i18n-key="subTitle">Select a Dashboard from menu to get started.</p>
```
```suggestion
                    <p data-i18n-key="subTitle"></p>
```

caponetto commented on dashbuilder/dashbuilder-runtime/src/main/java/org/dashbuilder/client/screens/view/UploadDashboardsScreenView.html:
```diff
@@ -0,0 +1,36 @@
+    <div class="blank-slate-pf-icon">
+        <span class="pficon pficon pficon-add-circle-o"></span>
+    </div>
+    <h1 data-i18n-key="title">Dashboards are not available</h1>
+    <p data-i18n-key="subTitle">Dashboards were not imported. You
+        can import them by uploading a file previously exported from
+        Business Central.</p>
+    <div class="blank-slate-pf-main-action">
+        <button class="btn btn-primary btn-lg" id="btnImport"
+            data-i18n-key="importButton">Import Dashboards</button>
+    </div>
```
```suggestion
    <h1 data-i18n-key="title"></h1>
    <p data-i18n-key="subTitle"></p>
    <div class="blank-slate-pf-main-action">
        <button class="btn btn-primary btn-lg" id="btnImport"
            data-i18n-key="importButton"></button>
    </div>
```

barboras7 commented:
Jenkins execute full downstream build

jesuino commented:
Thank you, @caponetto  - I just finished your last review, please see my latest commit.

Hello @barboras7 in case the build never finishes on time I have Runtime WAR here -> https://drive.google.com/file/d/1nnMFC4F9T-NnGmEsi1QC7BvT2uoyn7pS/view?usp=sharing

Just deploy it on widlfly 18 and login with an user with "admin" role, then import some dashboard exported from BC. It should run on root context and Kie Server dashboards are not supported. Once this PR is merged we will send the Kie Server one, which was built on top of this.

Thanks!

barboras7 commented:
Hi @jesuino,

I tried to reach out to Savitha today but I got no response. If she won't be able to look at the PR, I'll start working on it tomorrow. 

Anyway, I tried to run the .war file you provided and it seems that it started correctly. Anyway, I was not able to import any data. I am getting "Error uploading dashboards". I exported it from a business central I was using from Gradual export API pull request. Any idea why the import does not work?

jesuino commented:
Thank you for working on this, @barboras7 

Runtime should be available in root context: localhost:8080/. If it is in root context, then would you please tell me if you see any error in server.log or in the client console? Could you please also send by email the exported ZIP?

Thanks!

jesuino commented:
Hello @barboras7 

I am sorry, I just noticed that runtime was not configured by default to be on root. I made a new war with the default configure and built with FULL profile, which means that it is faster to load. 


https://drive.google.com/file/d/1nnMFC4F9T-NnGmEsi1QC7BvT2uoyn7pS/view?usp=sharing

jesuino commented:
I noticed I forgot GZIP filter, hence I added in my last commit. 

It is very important to have GZIP like we do in BC because it decreases the client payload. In Runtime case, a brief analysis shows the following:

* Without GZIP client paylog is almost 8mb
* With GZIP: 3mb

I updated the built WAR with GZIP version:

https://drive.google.com/file/d/1nnMFC4F9T-NnGmEsi1QC7BvT2uoyn7pS/view?usp=sharing

jesuino commented:
Jenkins execute full downstream build

jesuino commented:
Just FYI I found an issue which I will be working on soon

https://issues.redhat.com/browse/AF-2527

barboras7 commented:
Hi @jesuino,
following are some comments on the PR after my testing:

1. I am getting a warning in the wildfly log when starting the dashbuilder runtimes
`22:15:05,516 WARN  [org.jboss.weld.Validator] (MSC service thread 1-4) WELD-001471: Interceptor method createBaseDir defined on class org.dashbuilder.backend.resources.UploadResourceImpl is not defined according to the specification. It should not throw java.io.IOException, which is a checked exception.vat org.dashbuilder.backend.resources.UploadResourceImpl.createBaseDir(UploadResourceImpl.java:0)
`
2. When a page used in navigation is imported without a data set used in the page, there is an error that does not really look well (maybe some error message instead of showing runtime exception would be nice)
![Screenshot from 2020-05-25 22-03-39](https://user-images.githubusercontent.com/11525583/82842180-da406800-9ed8-11ea-9ef3-4cc032505300.png)
3. When there is no data imported into the dashbuilder runtime, it is still possible to click "Dashboards" and the empty menu opens, but there is nothing there (it's just a minor UX issue but maybe this could be disabled when there's no data)
![Screenshot from 2020-05-25 22-43-19](https://user-images.githubusercontent.com/11525583/82842329-5f2b8180-9ed9-11ea-8c5e-c92e5e19f394.png)
4. When you try to log in with a user without admin role, you get a "Forbidden" title without any option to navigate back (an error message would be better instead of letting the user to log in)
![Screenshot from 2020-05-25 22-46-41](https://user-images.githubusercontent.com/11525583/82842391-a0239600-9ed9-11ea-8910-507a3946f163.png)
5. I am not sure if the menu created in the dashbuilder runtimes module is user friendly (the pages used in the original menu of business central, newly created nav trees and its groups seem to be displayed at the same level) - although I am not sure how this could be solved, just mentioning that the distinction is not obvious
![Screenshot from 2020-05-25 23-00-22](https://user-images.githubusercontent.com/11525583/82842885-8aaf6b80-9edb-11ea-963c-848fbc17f717.png)
 


Also, one question:
Will the .war file with the dashbuilder runtime module be also included in the full downstream build on pull requests?

jesuino commented:
Hello Barbora,

Thanks for your review.

See my responses underlying:

1) Fixed
2) I did try to fix it, but some error messages may be coming from underlying services, hence it may include more technical messages. At least the duplicated "java.lang.Exception" I was able to remove because this specific error was written from a class I identified;
3) I am checking how to achieve it for this PR because soon we will support embedding these reports and them I will come with state concept: EMPTY, EMBEDDED and LOADED - I will switch the nav bar according to the state.
4) Fixed
5) The menu is following BC standard, which is the Mega Menu - and it allows only one menu level, hence this can't be changed ATM. Users, however, could make use of navigation elements to build deeper menus,

Regarding your question: Yes, it should be part of regular builds, including FDB. Later in this PR I noticed I didn't add the runtime to dashbuilder parent, but it is corrected now.

jesuino commented:
Jenkins execute full downstream build

barboras7 commented:
Thanks for the fixes, @jesuino! I tested also the `dashbuilder.runtime.import` property and it works. 

One small comment - the login with a user with correct credentials but without admin role redirects you to the "Not authorized" view. However, if you try to log in with a user that does not exist (random username and password), nothing obvious happens. On business central, you would get a "not authorized" error message. IMHO, it would be good to unify these. 

If the PR needs to be merged ASAP, I can just create a Jira for the above mentioned issue if we agree it needs to be fixed and we can merge the PR. :slightly_smiling_face:

ederign commented:
Please jenkins execute full downstream build

jesuino commented:
@barboras7 

Thank you so much for your review, really appreciate.

Would you please create a JIRA for this issue and I will work on it soon.  We must merge this to unblock the Kie Server PR

Thanks!

barboras7 commented:
The issue created here: **[AF-2537](https://issues.redhat.com/browse/AF-2537)**

sthundat commented:
Jira for no server disconnection error message: [AF-2538](https://issues.redhat.com/browse/AF-2538)
